### ISSUE TITLE
Create flask upstart script in puppet manifest

### ISSUE THREAD

**Issue by jeff1evesque** [jeff1evesque/machine-learning#1997]
We need to create `start_webserver.pp`.  This script will be responsible for installing flask, and defining the necessary [upstart](https://www.digitalocean.com/community/tutorials/the-upstart-event-system-what-it-is-and-how-to-use-it) script, to ensure that our flask server is running each time the Ubuntu Server has started.

The following need to be removed (if present):

- `flask` in `$packages_flask_pip   = ['flask', 'requests']` from `install_packages.pp`

**Comment by jeff1evesque** [comment]
After #2001 has been resolved, we will move corresponding python package installation from `install_packages.pp` to `start_webserver.pp`.

**Comment by jeff1evesque** [comment]
Since Ubuntu is [replacing](https://wiki.ubuntu.com/SystemdForUpstartUsers) [upstart](http://upstart.ubuntu.com/) with [systemd](https://wiki.ubuntu.com/systemd) scripts, we will implement the latter.

**Comment by jeff1evesque** [comment]
We need to create the following upstart script:

```conf
#!upstart
description 'Start Flask Server'

## start job defined in this file after system services, and
#      processes have already loaded (to prevent conflict).
#
#  @[2345] represents all configuration states with general linux access, and networking
start on runlevel [2345]
```

**Comment by jeff1evesque** [comment]
4f4f564: we forgot that redhat based distros may have a different format for upstart scripts.

**Comment by jeff1evesque** [comment]
We will make a minor change to the docblock.

**Comment by jeff1evesque** [comment]
We need to fix a minor language typo.

**Comment by jeff1evesque** [comment]
Our current upstart script does not automatically start on successive start-up.

**Comment by jeff1evesque** [comment]
aae5e9b: we've renamed `start_flask.conf` to `flask.conf`, since the upstart script should have the ability to `stop`, as well as `start` the corresponding service.

**Comment by jeff1evesque** [comment]
Since our custom logs defined in `flask.conf` are not executed properly, we need to check `/var/log/upstart`, and compare the logs for the following cases (in sequence):

- initial `vagrant up`
- successive `vagrant up` (second instance suffices)
- manual `sudo service flask start`
- manual `sudo reboot`

**Comment by jeff1evesque** [comment]
After initial `vagrant up`, we have the following upstart logs in `/var/log/upstart`:

```bash
vagrant@vagrant-ubuntu-trusty-64:/var/log/upstart$ ls
console-setup.log     network-interface-eth0.log      rsyslog.log
container-detect.log  pollinate.log                   statd.log
cryptdisks.log        procps-static-network-up.log    systemd-logind.log
gssd.log              procps-virtual-filesystems.log  ttyS0.log
```

This means, our flask service did not generate a default log.

**Comment by jeff1evesque** [comment]
On the successive `vagrant halt`, followed by `vagrant up`, we have the `/var/log/flask.log` generated, with the following content:

```
/bin/sh: 1: /bin/sh: cannot create /vagrant/log/flask_server.log: Directory nonexistent
```

**Comment by jeff1evesque** [comment]
We may need to add the following to start_webserver.pp:

pre-start script
    [ "$MOUNTPOINT" == "/vagrant" ] || stop
end script

**Comment by jeff1evesque** [comment]
e67c5d0: the upstart script properly starts as initial `vagrant up`, as displayed in `/var/log/upstart/flask.log`:

>*Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
* Restarting with stat

On successive `vagrant halt`, followed by `vagrant up`, we have the following in `/var/log/upstart/flask.log`:

>/proc/self/fd/9: 2: /proc/self/fd9: cannot create /vagrant/log/flask_server.log: Protocol error
*Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
* Restarting with stat

**Comment by jeff1evesque** [comment]
80001ca: after numerous `vagrant halt`, followed by `vagrant up`, we notice that our `/etc/init/flask.conf` correctly creates a background service:

```bash
vagrant@vagrant-ubuntu-trusty-64:~$ ps aux | grep python
root      1818  3.2  9.0 316912 45328 ?        Ss   09:16   0:00 python /vagrant
/app.py
root      1877  6.0  9.1 391032 45864 ?        Sl   09:16   0:01 /usr/bin/python
 /vagrant/app.py
vagrant   2123  0.0  0.1  10460   936 pts/1    S+   09:16   0:00 grep --color=au
to python
```

**Note:** since the system has been rebooted more than once, we notice that previous service processes are properly terminated.

**Comment by jeff1evesque** [comment]
If we clear the content of `/var/log/upstart/flask.conf`, the proceed with `vagrant halt`, then `vagrant up`, we get the following logs within `/var/log/upstart/flask.conf`:

```
/bin/sh: 1: /bin/sh: cannot create /vagrant/log/flask_server.log
: Protocol error
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
```

This means our `post-stop script` is unable to log when our upstart service terminate.

**Comment by jeff1evesque** [comment]
We need to define the following:

- `setuid`
- `setgid`

Otherwise, we get the following within `/var/log/upstart/flask.log` during `vagrant halt`:

```
/proc/self/fd/9: 2: /proc/self/fd/9: /vagrant: permission denied
Exception in thread Thread-1 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 763, in run
  File "/usr/local/lib/python2.7/dist-packages/werkzeug/serving.py", line 603 in inner
  File "/usr/local/lib/python2.7/dist-packages/werkzeug/serving.py", line 462 in serve_forever
  File "/usr/lib/python2.7/SocketServer.py", line 241, in serve_forever
  File "/usr/lib/python2.7/threading.py", line 585, in set
  File "/usr/lib/python2.7/threading.py", line 406, in notifyAll
<type 'exceptions.TypeError'>: 'NoneType' object is not callable
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
```

**Comment by jeff1evesque** [comment]
2d93aed: after numerous `vagrant halt`, followed by `vagrant up` we get the following within `/var/log/upstart/flask.log`:

```
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
```

Similarly, our `/vagrant/log/flask_server.log` contains the following:

```
[Sat Jul 11 17:47:19 UTC 2015] flask server starting
[Sat Jul 11 17:52:11 UTC 2015] flask server stopping
[Sat Jul 11 17:54:00 UTC 2015] flask server starting
[Sat Jul 11 17:54:46 UTC 2015] flask server stopping
[Sat Jul 11 17:56:42 UTC 2015] flask server starting
```

**Comment by jeff1evesque** [comment]
After `vagrant destroy`, `vagrant up`, `vagrant halt`, and `vagrant up` we notice the following error:

```bash
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
/proc/self/fd/9: 2: [: =: unexpected operator
```

The reason for this, is because we are running a windows 7 host, and when `setup_webserver.pp` creates `/etc/init/flask.conf`, it does not convert windows line endings within the linux guest machine.  If we implement the following within the Ubuntu virtual machine, our problems are eliminated:

```bash
sudo apt-get install dos2unix
sudo dos2unix /etc/init/flask.conf
```

**Comment by jeff1evesque** [comment]
We need to remove an unnecessary `notify => Exec['dos2unix-line-endings'],`.

**Comment by jeff1evesque** [comment]
We need to ensure that our flask service is restarted if `/etc/init/flask.conf` is modified.  Specifically,`/etc/init/flask.conf` will be modified (if windows host machine) via `dos2unix` conversion (i.e. line endings), during initial `vagrant up`.  Without the latter proposed changes, the `pre-stop script` stanza will properly execute only after the second `vagrant up` implementation.

**Comment by jeff1evesque** [comment]
We need to determine if the following from [`start_webserver.pp`](https://github.com/jeff1evesque/machine-learning/blob/ad2c24c21feeb86f125d1714af2b53f6d46b3f9a/puppet/manifests/start_webserver.pp#L75-L82) has some syntax error:

```puppet
...
            notify => Service['flask'],
        }

        ## start webserver
        service {'flask':
            ensure => 'running',
            enable => 'true',
        }
...
```

The motivation for the above snippet, is because our generated `/etc/init/flask.conf` does not automatically run on initial build, nor successive `vagrant up`.  However, if we manually `vagrant up`, then run `sudo service flask start`, then our flask server, `app.py` is guaranteed to be running.

**Comment by jeff1evesque** [comment]
We need to adjust our flask upstart script by implementing another [workaround](https://github.com/jeff1evesque/machine-learning/issues/2046#issuecomment-133593016) upstart script in order to ensure the `vagrant-mounted` event.

**Comment by jeff1evesque** [comment]
The workaround has been implemented in #2046. However, we will close this issue, since the creation of the upstart script has been resolved.
